[Windows] The filename, directory name, or volume label syntax is incorrect for *.html report

I am trying to run test by spoon on windows 7 machine but unfortunately something is wrong with the path to *.html reports for each test case.

I use spoon runner in version - 1.0.5 I am getting such error (the test are running on emulator fine)

```
[SR.runTests] Executing instrumentation suite on 1 device(s).
Exception in thread "main" java.lang.RuntimeException: java.io.FileNotFoundException: spoon-output\logs\192.168.56.102:5555\simeTest.html (The filename, directory name,
        at com.squareup.spoon.html.HtmlRenderer.renderMustacheToFile(HtmlRenderer.java:150)
        at com.squareup.spoon.html.HtmlRenderer.generateLogHtml(HtmlRenderer.java:138)
        at com.squareup.spoon.html.HtmlRenderer.render(HtmlRenderer.java:58)
        at com.squareup.spoon.SpoonRunner.run(SpoonRunner.java:80)
        at com.squareup.spoon.SpoonRunner.main(SpoonRunner.java:376)
Caused by: java.io.FileNotFoundException: spoon-output\logs\192.168.56.102:5555\simpleTest.html (The filename, directory name, or volume label syntax is incorrect)
        at java.io.FileOutputStream.open(Native Method)
        at java.io.FileOutputStream.<init>(Unknown Source)
        at java.io.FileOutputStream.<init>(Unknown Source)
        at java.io.FileWriter.<init>(Unknown Source)
        at com.squareup.spoon.html.HtmlRenderer.renderMustacheToFile(HtmlRenderer.java:147)
        ... 4 more
```

What I can see is that html files for each test case are under 'test' directory not 'logs' but also without 192... directory so it can be also a problem when we try to run tests on few devices (we are using genymotion for this).

When I display 'index.html' I can see such error:

```
com.android.ddmlib.SyncException: Target directory doesn't exist.
    at com.android.ddmlib.SyncService.pull(SyncService.java:250)
    at com.squareup.spoon.SpoonDeviceRunner.run(SpoonDeviceRunner.java:215)
    at com.squareup.spoon.SpoonRunner.runTests(SpoonRunner.java:111)
    at com.squareup.spoon.SpoonRunner.run(SpoonRunner.java:78)
    at com.squareup.spoon.SpoonRunner.main(SpoonRunner.java:376)
```

--- a/spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java
+++ b/spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java
@@ -40,6 +40,7 @@ public final class SpoonDeviceRunner {
   private static final String FILE_RESULT = "result.json";
   static final String TEMP_DIR = "work";
   static final String JUNIT_DIR = "junit-reports";
+  static final String IMAGE_DIR = "image";
 
   private final File sdk;
   private final File apk;
@@ -48,12 +49,12 @@ public final class SpoonDeviceRunner {
   private final boolean debug;
   private final boolean noAnimations;
   private final int adbTimeout;
-  private final File output;
   private final String className;
   private final String methodName;
   private final IRemoteAndroidTestRunner.TestSize testSize;
   private final File work;
   private final File junitReport;
+  private final File imageDir;
   private final String classpath;
   private final SpoonInstrumentationInfo instrumentationInfo;
 
@@ -74,8 +75,9 @@ public final class SpoonDeviceRunner {
    *        {@code className}.
    */
   SpoonDeviceRunner(File sdk, File apk, File testApk, File output, String serial, boolean debug,
-      boolean noAnimations, int adbTimeout, String classpath, SpoonInstrumentationInfo instrumentationInfo,
-      String className, String methodName, IRemoteAndroidTestRunner.TestSize testSize) {
+      boolean noAnimations, int adbTimeout, String classpath,
+      SpoonInstrumentationInfo instrumentationInfo, String className, String methodName,
+      IRemoteAndroidTestRunner.TestSize testSize) {
     this.sdk = sdk;
     this.apk = apk;
     this.testApk = testApk;
@@ -83,14 +85,16 @@ public final class SpoonDeviceRunner {
     this.debug = debug;
     this.noAnimations = noAnimations;
     this.adbTimeout = adbTimeout;
-    this.output = output;
     this.className = className;
     this.methodName = methodName;
     this.testSize = testSize;
-    this.work = FileUtils.getFile(output, TEMP_DIR, serial);
-    this.junitReport = FileUtils.getFile(output, JUNIT_DIR, serial + ".xml");
     this.classpath = classpath;
     this.instrumentationInfo = instrumentationInfo;
+
+    serial = SpoonUtils.sanitizeSerial(serial);
+    this.work = FileUtils.getFile(output, TEMP_DIR, serial);
+    this.junitReport = FileUtils.getFile(output, JUNIT_DIR, serial + ".xml");
+    this.imageDir = FileUtils.getFile(output, IMAGE_DIR, serial);
   }
 
   /** Serialize to disk and start {@link #main(String...)} in another process. */
@@ -223,7 +227,6 @@ public final class SpoonDeviceRunner {
 
       File screenshotDir = new File(work, dirName);
       if (screenshotDir.exists()) {
-        File imageDir = FileUtils.getFile(output, "image", serial);
         imageDir.mkdirs();
 
         // Move all children of the screenshot directory into the image folder.

--- a/spoon-runner/src/main/java/com/squareup/spoon/SpoonRunner.java
+++ b/spoon-runner/src/main/java/com/squareup/spoon/SpoonRunner.java
@@ -7,6 +7,7 @@ import com.beust.jcommander.JCommander;
 import com.beust.jcommander.Parameter;
 import com.beust.jcommander.ParameterException;
 import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableSet;
 import com.squareup.spoon.html.HtmlRenderer;
 import java.io.File;
 import java.io.IOException;
@@ -24,7 +25,6 @@ import static com.squareup.spoon.SpoonInstrumentationInfo.parseFromFile;
 import static com.squareup.spoon.SpoonLogger.logDebug;
 import static com.squareup.spoon.SpoonLogger.logInfo;
 import static java.util.Collections.synchronizedSet;
-import static java.util.Collections.unmodifiableSet;
 
 /** Represents a collection of devices and the test configuration to be executed. */
 public final class SpoonRunner {
@@ -47,8 +47,9 @@ public final class SpoonRunner {
   private final IRemoteAndroidTestRunner.TestSize testSize;
 
   private SpoonRunner(String title, File androidSdk, File applicationApk, File instrumentationApk,
-      File output, boolean debug, boolean noAnimations, int adbTimeout, Set<String> serials, String classpath,
-      String className, String methodName, IRemoteAndroidTestRunner.TestSize testSize) {
+      File output, boolean debug, boolean noAnimations, int adbTimeout, Set<String> serials,
+      String classpath, String className, String methodName,
+      IRemoteAndroidTestRunner.TestSize testSize) {
     this.title = title;
     this.androidSdk = androidSdk;
     this.applicationApk = applicationApk;
@@ -61,13 +62,7 @@ public final class SpoonRunner {
     this.methodName = methodName;
     this.classpath = classpath;
     this.testSize = testSize;
-
-    // Sanitize the serials for use on the filesystem as a folder name.
-    Set<String> serialsCopy = new LinkedHashSet<String>(serials.size());
-    for (String serial : serials) {
-      serialsCopy.add(SpoonUtils.sanitizeSerial(serial));
-    }
-    this.serials = unmodifiableSet(serialsCopy);
+    this.serials = ImmutableSet.copyOf(serials);
   }
 
   /**
@@ -124,13 +119,14 @@ public final class SpoonRunner {
     if (targetCount == 1) {
       // Since there is only one device just execute it synchronously in this process.
       String serial = serials.iterator().next();
+      String safeSerial = SpoonUtils.sanitizeSerial(serial);
       try {
         logDebug(debug, "[%s] Starting execution.", serial);
-        summary.addResult(serial, getTestRunner(serial, testInfo).run(adb));
+        summary.addResult(safeSerial, getTestRunner(serial, testInfo).run(adb));
       } catch (Exception e) {
         logDebug(debug, "[%s] Execution exception!", serial);
         e.printStackTrace(System.out);
-        summary.addResult(serial, new DeviceResult.Builder().addException(e).build());
+        summary.addResult(safeSerial, new DeviceResult.Builder().addException(e).build());
       } finally {
         logDebug(debug, "[%s] Execution done.", serial);
       }
@@ -139,13 +135,14 @@ public final class SpoonRunner {
       final CountDownLatch done = new CountDownLatch(targetCount);
       final Set<String> remaining = synchronizedSet(new HashSet<String>(serials));
       for (final String serial : serials) {
+        final String safeSerial = SpoonUtils.sanitizeSerial(serial);
         logDebug(debug, "[%s] Starting execution.", serial);
         new Thread(new Runnable() {
           @Override public void run() {
             try {
-              summary.addResult(serial, getTestRunner(serial, testInfo).runInNewProcess());
+              summary.addResult(safeSerial, getTestRunner(serial, testInfo).runInNewProcess());
             } catch (Exception e) {
-              summary.addResult(serial, new DeviceResult.Builder().addException(e).build());
+              summary.addResult(safeSerial, new DeviceResult.Builder().addException(e).build());
             } finally {
               done.countDown();
               remaining.remove(serial);
@@ -367,7 +364,8 @@ public final class SpoonRunner {
     @Parameter(names = { "--no-animations" }, description = "Disable animated gif generation")
     public boolean noAnimations;
 
-    @Parameter(names = { "--adb-timeout" }, description = "Set maximum execution time per test in seconds (10min default)")
+    @Parameter(names = { "--adb-timeout" },
+        description = "Set maximum execution time per test in seconds (10min default)")
     public int adbTimeoutSeconds = DEFAULT_ADB_TIMEOUT;
 
     @Parameter(names = { "--debug" }, hidden = true)

